### PR TITLE
Add internal solution for Barn Tree (usaco-1254)

### PR DESCRIPTION
**Pull request by aaggupta07**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/silver/usaco-1254.mdx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define f first
+#define s second
+#define MAX_N 200'001
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:30]: put in constant

File: solutions/silver/usaco-1254.mdx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+</CPPSection>
+
+<PySection>
+          
+</PySection>
+
+<JavaSection>
+
+</JavaSection>
+
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:120]:
```suggestion
```

File: solutions/silver/usaco-1254.mdx (hunk 1 above)
Comment (aaggupta07) [reply solutions/silver/usaco-1254.mdx]: fixed

File: solutions/silver/usaco-1254.mdx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define f first
+#define s second
+constexpr int MAX_N = 200'001;
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:30]:
```suggestion
const int MAX_N = 200001;
```
also make it 2e5 if possible (0-index the nodes)

File: solutions/silver/usaco-1254.mdx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+ll n, cnt;
+
+// Move hay bales and generate orders
+ll dfs(ll node, ll prev) {
+    ll balesToMove = bales[node];
+
+    for(ll i: adj[node]) {
+        if(i == prev) continue;
+        ll childBales = dfs(i, node);
+        balesToMove += childBales;
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:50]: tabs

File: solutions/silver/usaco-1254.mdx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        }
+        // if childBales is 0, we don't need to move any orders
+    }
+    return balesToMove;
+}
+
+/*
+For each barn, we print out all outgoing orders of the barns it receives an order from.
+This ensures that we always have a valid ordering of orders.
+*/
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:72]:
```suggestion
/*
 * For each barn, we print out all outgoing orders of the barns it receives an order from.
 * This ensures that we always have a valid ordering of orders.
*/
```

File: solutions/silver/usaco-1254.mdx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+void printOrders(ll node) {
+    visited[node] = true;
+    for(ll i: ordersFrom[node]) {
+        if(!visited[i]) printOrders(i);
+    }
+    for(auto& i: ordersTo[node]) cout << node + 1 << ' ' << i.f + 1 << ' ' << i.s << '\n';
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:78]: braces + over 80 chars

File: solutions/silver/usaco-1254.mdx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    return balesToMove;
+}
+
+/*
+For each barn, we print out all outgoing orders of the barns it receives an order from.
+This ensures that we always have a valid ordering of orders.
+*/
+void printOrders(ll node) {
+    visited[node] = true;
+    for(ll i: ordersFrom[node]) {
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:75]:
```suggestion
    for (ll i : ordersFrom[node]) {
```
similar for other control flow statements

File: solutions/silver/usaco-1254.mdx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        if(!visited[i]) printOrders(i);
+    }
+    for(auto& i: ordersTo[node]) cout << node + 1 << ' ' << i.f + 1 << ' ' << i.s << '\n';
+}
+
+
+signed main() {
+    ios::sync_with_stdio(false); cin.tie(nullptr);
+
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:84]:
```suggestion
int main() {
```

File: solutions/silver/usaco-1254.mdx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        adj[e-1].emplace_back(u - 1);
+    }
+
+    dfs(0, -1);
+    cout << cnt << '\n';
+
+    for(ll i = 0; i < n; ++i) {
+        if(!visited[i]) printOrders(i);
+    }
+    return 0;
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:107]:
```suggestion
```

File: solutions/silver/usaco-1254.mdx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    for(ll i = 0; i < n; ++i) bales[i] -= avg;
+
+    for(ll i = 1; i < n; ++i) {
+        ll u, e; cin >> u >> e;
+        adj[u-1].emplace_back(e - 1);
+        adj[e-1].emplace_back(u - 1);
+    }
+
+    dfs(0, -1);
+    cout << cnt << '\n';
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:102]:
```suggestion
    cout << cnt << endl;
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/silver/usaco-1254.mdx (hunk 4 above)
Comment (aaggupta07) [reply solutions/silver/usaco-1254.mdx:50]: they should be already...md collapses tabs into several spaces i think?

Comment (SansPapyrus683) [reply solutions/silver/usaco-1254.mdx:50]: if you highlight it you'll see that it's still spaces

File: solutions/silver/usaco-1254.mdx [hunk 11]
```diff
@@ -27,7 +27,7 @@ using namespace std;
 
 #define f first
 #define s second
-constexpr int MAX_N = 200'001;
+const int MAX_N = 2e5;
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:30]: you sure the code still works?

File: solutions/silver/usaco-1254.mdx (hunk 4 above)
Comment (aaggupta07) [reply solutions/silver/usaco-1254.mdx]: mb, should be fixed...think my editor was in space indent by accident.

File: solutions/silver/usaco-1254.mdx (hunk 11 above)
Comment (aaggupta07) [reply solutions/silver/usaco-1254.mdx:30]: yep, submitted to usaco to be sure, it works.

File: solutions/silver/usaco-1254.mdx [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define f first
+#define s second
+const int MAX_N = 2e5;
+
+using ll = long long;
+using pi = pair<ll, ll>;
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:32]: are long longs necessary?

File: solutions/silver/usaco-1254.mdx [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+vector<ll> ordersFrom[MAX_N]; // Stores for each barn the barns that it receives an order from
+vector<pi> ordersTo[MAX_N]; // Stores for each barn its outgoing orders
+
+ll bales[MAX_N];
+bool visited[MAX_N];
+ll n, cnt;
+
+// Move hay bales and generate orders
+ll dfs(ll node, ll prev) {
+	ll balesToMove = bales[node];
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:44]: c++ is snake_case in the guide

File: solutions/silver/usaco-1254.mdx [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+#define f first
+#define s second
+const int MAX_N = 2e5;
+
+using ll = long long;
+using pi = pair<ll, ll>;
+
+vector<ll> adj[MAX_N];
+vector<ll> ordersFrom[MAX_N]; // Stores for each barn the barns that it receives an order from
+vector<pi> ordersTo[MAX_N]; // Stores for each barn its outgoing orders
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:36]:
```suggestion
vector<ll> ordersFrom[MAX_N];  // Stores for each barn the barns that it receives an order from
vector<pi> ordersTo[MAX_N];  // Stores for each barn its outgoing orders
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/silver/usaco-1254.mdx (hunk 12 above)
Comment (aaggupta07) [reply solutions/silver/usaco-1254.mdx:32]: changed all the ones I can safely change to int in most recent commit, this problem has pretty atrocious overflow, every case besides sample will overflow w/out long longs.

Comment (SansPapyrus683) [reply solutions/silver/usaco-1254.mdx:32]: thanks
also would like it if you changed `pi` to `pll`

Comment (aaggupta07) [reply solutions/silver/usaco-1254.mdx]: done

File: solutions/silver/usaco-1254.mdx [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+using ll = long long;
+using pll = pair<ll, ll>;
+
+vector<int> adj[MAX_N];
+vector<int> orders_from[MAX_N];  // Stores for each barn the barns that it receives an order from
+vector<pll> orders_to[MAX_N];  // Stores for each barn its outgoing orders
+
+int bales[MAX_N];
+bool visited[MAX_N];
+int n, cnt;
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:40]: add a comment indicating what `cnt` is

**Review by SansPapyrus683 — APPROVED** [review]
yeah we're basically done

File: solutions/silver/usaco-1254.mdx [hunk 16]
```diff
@@ -37,7 +37,7 @@ vector<pll> orders_to[MAX_N];  // Stores for each barn its outgoing orders
 
 int bales[MAX_N];
 bool visited[MAX_N];
-int n, cnt;
+int n, cnt;  // cnt is the number of total orders we need to make (in an optimal solution)
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1254.mdx:40]: ok in this case just name cnt to `order_num`
also declare it separately from `n`
sorry for the last-minute stuff, i swear this is iit

Comment (aaggupta07) [reply solutions/silver/usaco-1254.mdx]: np, should be good now

**Review by ryanchou-dev — APPROVED** [review]

**Pull request merged**